File: NASA_JulesKouatchou/Problem2/matMult_loops.c
/*
 **********************************************************
 Given two nxn matrices A and B, we perform:
       C = A x B
 **********************************************************
*/

#include <sys/types.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>



int main( int argc, char *argv[] )
{
        double **A, **B, **C;
	int i, j, k;
	int dim;
        clock_t start, finish; 
	double maxr;

        // ---------------------------------------
        // Get the dimension from the command line
        // ---------------------------------------
        dim = atoi(argv[1]);

        // -------------------------------------
        // Allocate and initialize the matrices
        // -------------------------------------

        A = malloc(sizeof(double *) * dim);
        B = malloc(sizeof(double *) * dim);
        C = malloc(sizeof(double *) * dim);
        for (i = 0; i < dim; i++) {
            A[i] = malloc(sizeof(double) * dim);
            B[i] = malloc(sizeof(double) * dim);
            C[i] = malloc(sizeof(double) * dim);
        }

	srand(time(NULL));
	maxr = (double)RAND_MAX;
	for (i = 0; i < dim; i++) {
            for (j = 0; j < dim; j++) {
		A[i][j] = rand()/maxr;
		B[i][j] = rand()/maxr;
            }
        }

        // ---------------------------------
        // Compute the matrix multiplication
        // ---------------------------------
	start = clock();

	for (i = 0; i < dim; i++)
	{
		for (j = 0; j < dim; j++)
			C[i][j] = 0.;
		for (k = 0; k < dim; k++)
			for (j = 0; j < dim; j++)
				C[i][j] += A[i][k]*B[k][j];
	}
	finish = clock();

	printf("time for C(%d,%d) = A(%d,%d) B(%d,%d) is %lf s\n", dim, dim, dim, dim, dim, dim, (double) (finish - start)/CLOCKS_PER_SEC);	
	return 0;
}

File: NASA_JulesKouatchou/Problem2/matMult_loops.f90
! **********************************************************
! Given two nxn matrices A and B, we perform:
!       C = A x B
! **********************************************************

      program matMult_loops

      IMPLICIT NONE

      real :: r

      real start, finish

      integer dim
      double precision, allocatable :: A(:,:), B(:, :), C(:, :)

      integer :: i, j, k
      integer, parameter :: n = 1
      CHARACTER(len=50) :: arg(n)

      ! Get the dimension from the command line
      !----------------------------------------
      do i =1, n
         call getarg(i,arg(i))
      end do
      read(arg(1), * ) dim

      !-------------------------------------
      ! Allocate and initialize the matrices
      !-------------------------------------
      allocate(A(dim,dim))
      allocate(B(dim,dim))
      allocate(C(dim,dim))

      call srand(86456)
      do i = 1, dim
         do j = 1, dim
            call random_number(r)
            A(i, j) = r
            call random_number(r)
            B(i, j) = r
         enddo
      enddo

      !----------------------------------
      ! Perform the matrix multiplication
      !----------------------------------
      call cpu_time(start)
      do j = 1, dim
         do i = 1, dim
            C(i, j) = 0.
            do k = 1, dim
               C(i, j) = C(i, j) + A(i, k)*B(k, j)
            enddo
         enddo
      enddo
      call cpu_time(finish)

      print *,'Matrix multiplication with loops: time for C(',dim,',',dim,') = A(',dim,',',dim,') B(', &
     &   dim,',',dim,') is',finish - start,' s'

      end program matMult_loops
